### PR TITLE
Cortex Project Security Self-Assessment - Security Pals

### PR DESCRIPTION
**Pull request by heydc7**
Created and added the first draft of Cortex Project Security Self-Assessment. Please feel free to share your thoughts on the security self-assessment.

Contributors: Dhanraj Chavan, Raiya Haque, Abdul Alhazmi, Sushanth Ravipalli

### CODE REVIEW

**Comment by eddie-knight** [comment]
Hi there! I'm just getting started looking at your pull request, and I noticed the DCO check is failing.

You can look at the checks section of the PR (I believe it should always be below the last comment) and look for a red X highlighting the failed check. In this case, you can click `Details` for more information about how to get that check passing.
Screenshot 2023-12-08 at 8 35 18 AM

**Comment by heydc7** [comment]
Sure, I'll remove SBOM from this assessment. I have fixed the DCO check.

**Comment by heydc7** [comment]
Done. Removed SBOM ✅

File: assessments/projects/cortex/self-assessment.md [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+## Metadata
+
+A table at the top for quick reference information, later used for indexing.
+
+|   |  |
+| -- | -- |
+| Software | https://github.com/cortexproject/cortex |
+| Security Provider | No |
+| Languages | Go |
+| SBOM | `Not Applicable`  |
```
Comment (eddie-knight) [inline assessments/projects/cortex/self-assessment.md:41]:
```suggestion
| SBOM | [Built on release](https://github.com/cortexproject/cortex/releases/latest/download/sbom.tar.gz)  |
```

File: assessments/projects/cortex/self-assessment.md [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+The Self-assessment is the initial document for projects to begin thinking about the
+security of the project, determining gaps in their security, and preparing any security
+documentation for their users. This document is ideal for projects currently in the
+CNCF **sandbox** as well as projects that are looking to receive a joint assessment and
+currently in CNCF **incubation**.
+
+For a detailed guide with step-by-step discussion and examples, check out the free 
+Express Learning course provided by Linux Foundation Training & Certification: 
+[Security Assessments for Open Source Projects](https://training.linuxfoundation.org/express-learning/security-self-assessments-for-open-source-projects-lfel1005/).
+
```
Comment (eddie-knight) [inline assessments/projects/cortex/self-assessment.md:11]: Please ensure that no template text like this is included here or in the rest of the document

**Comment by heydc7** [comment]
Ok, we will do it.

**Comment by heydc7** [comment]
Done ✅

File: assessments/projects/cortex/sbom.json
Comment (ragashreeshekar) [inline assessments/projects/cortex/sbom.json:1]: SBOM as stated previously needs to be removed. Kindly delete this from the PR and if publicly released by the team, include a link in the self-assessment

File: assessments/projects/cortex/self-assessment.md [hunk 3]
```diff
@@ -0,0 +1,261 @@
+# Cortex - Self-assessment
+
+The Self-assessment is the initial document for Cortex to begin thinking about the security of the project, determining gaps in its security, and preparing any security documentation for their users.
+
+Authors: Dhanraj Chavan @heydc7 Raiya Haque @raiyahaque Abdul Alhazmi @Abkzmii Sushanth Rabipalli sushanth3120
+Contributors/Reviewers: Pranava Kumar Vemula (@Rana-KV)
+
+# Self-assessment outline
```
Comment (ragashreeshekar) [inline assessments/projects/cortex/self-assessment.md:8]: All web content should only have a single H1 entry (#)

File: assessments/projects/cortex/self-assessment.md [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+* [Security functions and features](#security-functions-and-features)
+* [Project compliance](#project-compliance)
+* [Secure development practices](#secure-development-practices)
+* [Security issue resolution](#security-issue-resolution)
+* [Appendix](#appendix)
+* [Action Items](#action-items)
+
+## Metadata
+
+A table at the top for quick reference information, later used for indexing.
```
Comment (ragashreeshekar) [inline assessments/projects/cortex/self-assessment.md:30]: Please remove any generic template content such as this one from the self-assessment

File: assessments/projects/cortex/self-assessment.md [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+- **Threat:** Unauthorized escalation of user privileges within Cortex.
+- **Application to Cortex:** Not applicable in Cortex as there is no concept of a superuser or admin user.
+- **Mitigation:** Ensure adherence to Kubernetes access controls.
+
+### Additional Considerations:
+- **Alertmanager Security:** Address security threats around Alertmanager, which have already been mitigated. Focus on ensuring these mitigations remain effective.
+
+## Security issue resolution
+
+#### Responsible Disclosures Process:
```
Comment (ragashreeshekar) [inline assessments/projects/cortex/self-assessment.md:223]: Can you please provide reference to this process? If there are any triage process, remediation timelines, etc. kindly include them here

File: assessments/projects/cortex/self-assessment.md [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+- **Thanos:** A CNCF Incubating project known for its highly available Prometheus setup with long-term storage capabilities.
+- **Cortex:** Distinguished by its horizontally scalable architecture, high availability, multi-tenant support, and focus on long-term storage for Prometheus
+
+## Action Items
+- Engaged in collaborative self-assessment discussions with Project Maintainers on [Slack](https://cloud-native.slack.com/archives/C0682GTKK4K/p1701568106182349) and [GitHub Issue](https://github.com/cortexproject/cortex/issues/5692) to align goals and address feedback 
+- Implemented SBOM generation using cyclonedx-gomod, providing a detailed inventory of Go module dependencies.
+- Conducted a comprehensive update of the project's threat model with respect to Cortex Maintainer's Response.
+- Revised and expanded the project documentation's appendix section, incorporating additional materials to enhance overall understanding and documentation.
+- Ensured that all relevant links associated with the development pipeline and communication channels are up-to-date.
+
```
Comment (ragashreeshekar) [inline assessments/projects/cortex/self-assessment.md:261]: This section is meant to serve as action items for the project maintainers to improve the security posture of the project. Please include content that is relevant in this context

**Review by ragashreeshekar — CHANGES_REQUESTED** [review]
Thanks for the PR @heydc7 and team, appreciate the efforts.
I have completed first pass of review. Please feel free to reach out here or on slack for any questions and clarifications.

Along with addressing the comments, kindly update the PR branch with the latest content in the repo as this branch is out-of-date with the base branch.

**Comment by heydc7** [comment]
Hi @ragashreeshekar 
Can you please guide me on updating the PR branch with the latest content in the repo without losing my previous commits?

**Comment by heydc7** [comment]
PR Branch updated

**Comment by heydc7** [comment]
All changes are done ✅

File: assessments/projects/cortex/self-assessment.md [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+* [Action Items](#action-items)
+
+## Metadata
+
+|   |  |
+| -- | -- |
+| Software | https://github.com/cortexproject/cortex |
+| Security Provider | No |
+| Languages | Go |
+| SBOM | `Not Applicable` |
```
Comment (eddie-knight) [inline assessments/projects/cortex/self-assessment.md:36]:
```suggestion
| SBOM | https://github.com/cortexproject/cortex/releases/latest/download/sbom.tar.gz |
```

Comment (eddie-knight) [reply assessments/projects/cortex/self-assessment.md:36]: While Cortex isn't automatically generating the SBOM at build time (ideal), they do have a policy they appear to be following to manually generate it at release time using Kubernetes BOM.

https://github.com/cortexproject/cortex/blob/master/RELEASE.md#sign-the-release-artifacts-and-generate-sbom

File: assessments/projects/cortex/self-assessment.md [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+### Case Studies
+Detailed case studies or specific real-world use cases for Cortex were not found. However, Cortex is designed for large-scale cloud-native environments, suggesting its use in extensive monitoring and analysis of time-series data, showcasing its scalability, long-term storage, and multi-tenancy capabilities.
+
+### Related Projects / Vendors
+Comparing Cortex with similar projects like Thanos reveals key differences:
+- **Thanos:** A CNCF Incubating project known for its highly available Prometheus setup with long-term storage capabilities.
+- **Cortex:** Distinguished by its horizontally scalable architecture, high availability, multi-tenant support, and focus on long-term storage for Prometheus
+
+## Action Items
+- Suggested to create SBOM(Software Bill of Materials) for Cortex project: [Issue](https://github.com/cortexproject/cortex/issues/5692)
```
Comment (eddie-knight) [inline assessments/projects/cortex/self-assessment.md:258]: Cortex already has an SBOM being created (as noted above) and as a secondary point, the review team is now asking that action items or recommendations not be included in the security self-assessment.

Comment (heydc7) [reply assessments/projects/cortex/self-assessment.md:258]: Ok, Should I remove the Action Items section?

File: assessments/projects/cortex/self-assessment.md (hunk 7 above)
Comment (heydc7) [reply assessments/projects/cortex/self-assessment.md:36]: OK Will add.

File: assessments/projects/cortex/self-assessment.md (hunk 8 above)
Comment (heydc7) [reply assessments/projects/cortex/self-assessment.md]: Removed action items from assessment.

File: assessments/projects/cortex/self-assessment.md [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+- **Application to Cortex:** A well-configured Cortex system is resilient to DoS attacks.
+- **Mitigation:** Implement rate limiting and series per tenant limits to prevent DoS attacks.
+
+### Elevation of Privilege:
+- **Threat:** Unauthorized escalation of user privileges within Cortex.
+- **Application to Cortex:** Not applicable in Cortex as there is no concept of a superuser or admin user.
+- **Mitigation:** Ensure adherence to Kubernetes access controls.
+
+### Additional Considerations:
+- **Alertmanager Security:** Address security threats around Alertmanager, which have already been mitigated. Focus on ensuring these mitigations remain effective.
```
Comment (eddie-knight) [inline assessments/projects/cortex/self-assessment.md:216]: Following discussions among the review team, we are asking that all STRIDE content be moved to a secondary document within this directory. Please extract this content and add a link to the new file in the appendix.

Comment (heydc7) [reply assessments/projects/cortex/self-assessment.md]: Done ✅

File: assessments/projects/cortex/self-assessment.md [hunk 10]
```diff
@@ -0,0 +1,220 @@
+# Cortex - Self-assessment
+
+The Self-assessment is the initial document for Cortex to begin thinking about the security of the project, determining gaps in its security, and preparing any security documentation for their users.
```
Comment (torinvdb) [inline assessments/projects/cortex/self-assessment.md:3]:
```suggestion
# Cortex - Self-assessment

This assessment was created by community members as part of the [Security Pals](https://github.com/cncf/tag-security/issues/1102) process and is currently pending changes from the maintainer team.

The Self-assessment is the initial document for Cortex to begin thinking about the security of the project, determining gaps in its security, and preparing any security documentation for their users.
```
@ragashreeshekar

File: assessments/projects/cortex/self-assessment.md (hunk 6 above)
Comment (ragashreeshekar) [inline assessments/projects/cortex/self-assessment.md]:
```suggestion
```

File: assessments/projects/cortex/self-assessment.md (hunk 4 above)
Comment (ragashreeshekar) [inline assessments/projects/cortex/self-assessment.md]:
```suggestion
```

File: assessments/projects/cortex/self-assessment.md [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+- Cortex aims to run across a cluster of machines, handling metrics from multiple Prometheus servers and supporting globally aggregated queries.
+- In a cluster setup, Cortex replicates data between machines to ensure high availability in the face of machine failures.
+
+### Non-goals
+- Cortex does not aim to serve as a general-purpose database for various data types, focusing specifically on Prometheus metrics.
+- Cortex may not prioritize real-time processing or streaming of data, as its primary function is long-term, durable storage of metric data.
+
+## Self-assessment use
+
+This self-assessment is created by the Cortex team to perform an internal analysis of Cortex's security. It is not intended to provide a security audit of Cortex or function as an independent assessment or attestation of Cortex's security health.
```
Comment (ragashreeshekar) [inline assessments/projects/cortex/self-assessment.md:131]:
```suggestion
```

**Review by ragashreeshekar — APPROVED** [review]

**Review by eddie-knight — APPROVED** [review]

**Pull request merged**